### ISSUE TITLE
oneTimeTearDown()/tearDown() is being called an additional time at the end of execution

### ISSUE THREAD

**Issue by vtereso** [kward/shunit2#112]
I have been pulling master and had functions that were executing within these hooks and saw what seems to be some unintended behavior. Example file as show below:
```
testSomething() {
    assertTrue "true" "true"
}

setUp() {
  echo "setup"
}
tearDown() {
  echo "Tear down"
}
oneTimeTearDown() {
  echo "One time tear down"
}
. shunit2
```

Output:
```
setup
testSomething
Tear down
One time tear down

Ran 1 test.

OK
Tear down
One time tear down
```

**Comment by vtereso** [comment]
Looking into it, I only see warnings being raised if these functions do not exit successfully, but do not see why they would be called at the end of execution. PR for consideration: https://github.com/kward/shunit2/pull/113

**Comment by cbandy** [comment]
These are being called as part of the EXIT trap. Related to #108.

**Comment by vtereso** [comment]
Right, I wasn't familiarized with the entire topology. Looked things over and there is still a fundamental flaw since the exit code at the end of the script gets caught by the trap on successful runs so the 0 signal triggers _shunit_cleanup and the following block ```if command [ ${__shunit_clean} -eq ${SHUNIT_FALSE} ]; then
    # Ensure tear downs are only called once.
    __shunit_clean=${SHUNIT_TRUE}

    tearDown
    command [ $? -eq ${SHUNIT_TRUE} ] \
        || _shunit_warn "tearDown() returned non-zero return code."
    oneTimeTearDown
    command [ $? -eq ${SHUNIT_TRUE} ] \
        || _shunit_warn "oneTimeTearDown() returned non-zero return code."

    command rm -fr "${__shunit_tmpDir}"
  fi``` will trigger the tear downs again.

**Comment by jaspalch** [comment]
This can be used as sort of a hacky workaround:
```
tearDown() {
    #need this to suppress tearDown on script EXIT
    [[ "${_shunit_name_}" = 'EXIT' ]] && return 0
```

**Comment by swcline** [comment]
I tried adding the above code inside tearDown() but get the following error - is there something else that needs to be enabled or set?

in oneTimeSetUp
in setUp
testOne
validateTestOrdinals_test.sh[34]: _shunit_name_: parameter not set


the test script:

#! /bin/sh
# assume this test script is in a subdirectory under the shunit directory; a sibling of examples

# Load test helpers.
. ../shunit2_test_helpers

testOne() {
  assertTrue 0
}

testTwo() {
  assertTrue 0
}

setUp() {
  echo "in setUp"
}

tearDown() {
  echo "in tearDown"
}

oneTimeSetUp() {
  echo "in oneTimeSetUp"
}

oneTimeTearDown() {
echo "in oneTimeTearDown"
}

# Load and run shUnit2.
. ../shunit2

**Comment by swcline** [comment]
My workaround is to create a directory in oneTimeSetUp and wrap the tear down code inside a check for that directory inside oneTimeTearDown - second time around, directory is not there, so code is skipped

**Comment by ypoluektovich** [comment]
Yes.

**Comment by Bajger** [comment]
Yes, tested on release 2.1.8. My output looks like this:
```
Calling teardown...

Ran 4 tests.

OK
Calling teardown...
PharoLauncherCommonFunctions.sh: line 45: ./pharo-launcher.sh: No such file or directory
```

**Comment by swcline** [comment]
Here is a simple test that exhibits the issue

```sh
#! /bin/bash

# assume this test script is in a subdirectory under the shunit directory;
a sibling of examples



# Load test helpers.

. ../shunit2_test_helpers



testOne() {

  assertTrue 0

}



testTwo() {

  assertTrue 0

}



setUp() {

  echo "in setUp"

}



tearDown() {

  echo "in tearDown"

}



oneTimeSetUp() {

  #oneTimeTearDown gets called one extra time at exit, to prevent that add
and check a guard file

  #touch "${SHUNIT_TMPDIR}/guard.tmp"

  echo "in oneTimeSetUp"

}



oneTimeTearDown() {

  #oneTimeTearDown gets called one extra time at exit, to prevent that add
and check a guard file

  #if [ -f "${SHUNIT_TMPDIR}/guard.tmp" ]; then

  echo "in oneTimeTearDown"

  #fi

}



# Load and run shUnit2.

. ../shunit2
```

**Comment by cezanne** [comment]
I had the same problem. The following modification resolved my issue.

```
diff --git a/shunit2 b/shunit2
index 57a45da..ead9de6 100755
--- a/shunit2
+++ b/shunit2
@@ -1367,6 +1367,7 @@ _shunit_execSuite
 if ! oneTimeTearDown; then
   _shunit_fatal "oneTimeTearDown() returned non-zero return code."
 fi
+__shunit_clean=${SHUNIT_TRUE}
 
 # Generate a report summary.
 _shunit_generateReport
```

**Comment by larssb** [comment]
Could we please get one of the PR/suggestions to fix this merged into main on kward and get a new release? I just bumped into this and it took some troubleshooting to figure it out.

Would be great to avoid having shunit2 executing the `oneTimeTearDown` more than exactly once, unless it's a dirty exit.

Here's some of the suggestions I've found:

- https://github.com/kward/shunit2/pull/132
- https://github.com/cezanne/shunit2/commit/e09d0bfad4cb7b323b20a3ccfcd88d77d09ae48e

Thank you very much 👍

**Comment by larssb** [comment]
> My workaround is to create a directory in oneTimeSetUp and wrap the tear down code inside a check for that directory inside oneTimeTearDown - second time around, directory is not there, so code is skipped

An alternative to this is to set a global env. var in `Bash`.

E.g.

```bash
oneTimeSetUp()
{
    SOME_SETUP_LOGIC_HERE
    
    export avoidDoubleTearDownExecution="true"
}

....
....

oneTimeTearDown()
{
    if [[ "${avoidDoubleTeadDownExecution}" == "true" ]]; then
        SOME_TEARDOWN_LOGIC_HERE
        
        unset -v avoidDoubleTearDownExecution
    fi
}
```

The above makes the tear down, in clean cases, where what's in https://github.com/kward/shunit2/issues/108 is not needed, only run once.

Enjoy 🥇

**Comment by oschonrock** [comment]
same problem here, but I also find that EACH TEST function is executed  twice..
